### PR TITLE
New method "get" for the ConfigService class

### PR DESCRIPTION
**Pull request by jmborr**
Companion to PR #32506 

Description:
===========
The python API of Mantid's `ConfigService` class duck-types the python dictionary, but it doesn't have a `.get` method, e.g.
```python
from mantid.kernel import ConfigService
conf = ConfigService.Instance()
behavior = conf.get('algorithms.alias.deprecated', 'Warn')
```
The default value `Warn` above avoids some boiler plate code with a `try..except` block:
```python
try:
    behavior = conf['algorithms.alias.deprecated']
except AttributeError:
    behavior = 'Warn'
```

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by KedoKudo — COMMENTED** [review]
This is a very helpful feature to add to config service.
However, I believe there is one case missing here.
Here is a simple Python example
```python
>>> a = {'a': 1}
>>> a.get('a', 0)
1
>>> a.get('b', 0)
0
>>> a.get('b')
>>>
```
This PR enables the first two usages of `.get`, but the third one is still missing.
Is it possible to add an overloaded method such that querying a non-existing item will not trigger an error.

**Comment by jmborr** [comment]
> This is a very helpful feature to add to config service.
> However, I believe there is one case missing here.
> Here is a simple Python example
> 
> ```python
> >>> a = {'a': 1}
> >>> a.get('a', 0)
> 1
> >>> a.get('b', 0)
> 0
> >>> a.get('b')
> >>>
> ```
> 
> This PR enables the first two usages of `.get`, but the third one is still missing.
> Is it possible to add an overloaded method such that querying a non-existing item will not trigger an error.

You're right, I hadn't covered that case :blush: 
Can you give it a try now?

**Comment by KedoKudo** [comment]
Just give it another try, and here is what I got
```
from mantid.kernel import ConfigService
conf = ConfigService.Instance()
behavior = conf.get('algorithms.alias.deprecated', 'Warn')


conf.get('algorithms.alias.deprecated')
Out[2]: ''

conf.get('algorithms.alias.deprecated', 'a')
Out[3]: 'a'

conf.get('algorithms.alias.deprecated', None)
---------------------------------------------------------------------------
ArgumentError                             Traceback (most recent call last)
<ipython-input-4-473ef32dd9c1> in <module>()
----> 1 conf.get('algorithms.alias.deprecated', None)

ArgumentError: Python argument types in
    ConfigServiceImpl.get(ConfigServiceImpl, str, NoneType)
did not match C++ signature:
    get(Mantid::Kernel::ConfigServiceImpl {lvalue}, std::string self, std::string key)
    get(Mantid::Kernel::ConfigServiceImpl {lvalue} self, std::string key)

conf.get('algorithms.alias.deprecated', 1)
---------------------------------------------------------------------------
ArgumentError                             Traceback (most recent call last)
<ipython-input-5-b3a21d6d637e> in <module>()
----> 1 conf.get('algorithms.alias.deprecated', 1)

ArgumentError: Python argument types in
    ConfigServiceImpl.get(ConfigServiceImpl, str, int)
did not match C++ signature:
    get(Mantid::Kernel::ConfigServiceImpl {lvalue}, std::string self, std::string key)
    get(Mantid::Kernel::ConfigServiceImpl {lvalue} self, std::string key)
```

It seems like the default value provided must be string.  Is this a intended feature?  If so, I will approve this PR.  Otherwise, the type of the default value need to be more flexible (`PyObject *` perhaps?).

**Comment by jmborr** [comment]
the default value should be a string object, yes. This is to be consistent with `__getitem__` which returns an empty string if the key is not found, e.g. `config["property_not_found"]` returns and empty string.
I can write the documentation of this method to stress that the default value should be and object of type string

**Review by KedoKudo — DISMISSED** [review]
Looks good 👍

File: Framework/PythonInterface/mantid/kernel/src/Exports/ConfigService.cpp [hunk 1]
```diff
@@ -50,6 +50,16 @@ const InstrumentInfo &getInstrument(ConfigServiceImpl &self, const object &name
     return self.getInstrument(ExtractStdString(name)());
 }
 
+/// duck typing emulating dict.get method
+std::string getStringUsingCacheElseDefault(ConfigServiceImpl &self, const std::string &key,
+                                           const std::string &defaultValue) {
+  std::vector<std::string> keys = self.keys();
+  if (std::find(keys.begin(), keys.end(), key) != keys.end())
```
Comment (peterfpeterson) [inline Framework/PythonInterface/mantid/kernel/src/Exports/ConfigService.cpp:57]: Use the existing function
```suggestion
  if (hasProperty(key))
```

Comment (jmborr) [reply Framework/PythonInterface/mantid/kernel/src/Exports/ConfigService.cpp]: good catch! :smile:

**Review by KedoKudo — APPROVED** [review]

**Pull request merged**